Create counterparty.sol contract

0xmovses opened the pull request:
# Summary
Creates the `AtomicBridgeCounterparty.sol` contract, tests and updates necessary logic in `AtomicBridgeInitiator.sol`. I added a storage mapping of balances in the Initiator contract. This is so that when users move MOVETH -> WETH they are able to withdraw their associated balance. We could have used a seperate Vault type contract, but I decided to keep it as simple as possible with minimal change necessary. 

Because of this, counterpary needs to call a withdraw function on Initiator, where the balances mapping is held. This function should of course only be callable by that contract and not by anyone, For this I added the ownable patter. To keep the contract upgradable, I used `UpgradableOwnable`.

# Changelog
- Creates new contract
- Creates new test
- Uses `UpgradableOwnable.sol` openzeppelin lib for initiator
- Updates `AtomicBridgeInitiator.t.sol` to assert on the enum values. 
- Adds the solidity tests to Checks CI

# Testing
Please run the tests.
`cd protocol-units/bridge/contracts && forge test --fork-url https://ethereum-sepolia-rpc.publicnode.com`

cde8 commented on protocol-units/bridge/contracts/src/AtomicBridgeInitator.sol:
```diff
@@ -83,8 +97,19 @@ contract AtomicBridgeInitiator is IAtomicBridgeInitiator, Initializable {
     }
+
+    // Counterparty contract to withdraw WETH for originator
+    function withdrawWETH(address originator, uint256 amount) external {
```
Maybe call this the receipient, as this is how its called from the Counterpart contract.

0xmovses replied:
yep, better.

cde8 commented on protocol-units/bridge/contracts/src/AtomicBridgeInitator.sol:
```diff
@@ -46,12 +57,15 @@ contract AtomicBridgeInitiator is IAtomicBridgeInitiator, Initializable {
 
-        nonce++; //increment the nonce
+        // Update the balance of the originator 
+        balances[originator] += totalAmount;
```
Interesting approach to track balances per recipient. Why not just treat the whole thing as a pool?

0xmovses replied:
When you say treat it as a pool. What data type do you propose? For liquidity pools I've always seen mappings of addresses to amounts.

cde8 replied:
With pooled I mean, just throw all the ETH in one big bucket. Saves complexity and gas etc.

0xmovses replied:
hmm. Ok, lets give a try.

cde8 commented on protocol-units/bridge/contracts/src/AtomicBridgeInitator.sol:
```diff
@@ -83,8 +97,19 @@ contract AtomicBridgeInitiator is IAtomicBridgeInitiator, Initializable {
+    function withdrawWETH(address originator, uint256 amount) external {
+        require(msg.sender == counterpartyContract, "Caller is not the counterparty contract");
+        require(balances[originator] >= amount, "Insufficient balance");
+        balances[originator] -= amount;
```
The balances should exactly match, unless a recipient has multiple in flight. I would just treat it as a pool.

0xmovses replied:
I've made the check exactly `==`.

What data structure do you propose for the pool? I've seen mappings as a pretty common data structure for storing liquidity in a contract.

cde8 replied:
See comments above, I would just pool.

cde8 reviewed (changes requested):
Looking good, some small remarks

0xmovses commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,81 @@
+        details.state = MessageState.COMPLETED;
+
+        // Call withdrawWETH on AtomicBridgeInitiator to transfer funds to the recipient
+        atomicBridgeInitiator.withdrawWETH(details.recipient, details.amount);
```
Yes, this is discussed a little in the PR description. We could use a vault contract later down the line, but I went with the most minimal change initially. Balances are stored in the initiator contract.

0xmovses commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,81 @@
+
+        bytes32 computedHash = keccak256(abi.encodePacked(preImage));
+        if (computedHash != details.hashLock) revert InvalidSecret();
+        if (block.timestamp > details.timeLock) revert TimeLockNotExpired();
```
Ah yes, my mistake. Will update.

0xmovses commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,81 @@
+        atomicBridgeInitiator = AtomicBridgeInitiator(_atomicBridgeInitiator);
+    }
+
+    function lockBridgeTransferAssets(
+        bytes32 initiator,
+        bytes32 bridgeTransferId,
+        bytes32 hashLock,
+        uint256 timeLock,
+        address recipient,
+        uint256 amount
+    ) external returns (bool) {
```
I'll add access control onto this funciton.

0xmovses replied:
Added.

0xmovses commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,81 @@
+        MessageState state; 
+    }
+
+    AtomicBridgeInitiator public atomicBridgeInitiator;
```
As you discovered, we're calling a function on this contract to withdraw the WETH associated with the recipient address.

0xmovses commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,81 @@
+
+    function abortBridgeTransfer(bytes32 bridgeTransferId) external {
+        BridgeTransferDetails storage details = bridgeTransfers[bridgeTransferId];
+        if (details.state != MessageState.PENDING) revert BridgeTransferInvalid();
```
You mean error name, I assume, not event name. Sounds good.

0xmovses commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,81 @@
+        uint256 amount
+    ) external returns (bool) {
+        BridgeTransferDetails storage transfer = bridgeTransfers[bridgeTransferId];
+        if (transfer.recipient != address(0)) revert BridgeTransferInvalid();
```
This is input validation, it checks that the address is not `0x0000000000000000000000000000000000000000` as this would be a valid values for `address`. If the address is 0x, then we revert with the Error. Updated now to 
`        if (recipient != address(0)) revert BridgeTransferInvalid();` to check the arg, not the storage val.

0xmovses replied:
Sorry correction, it means, we are expecting the `transfer.recipient` to be `address(0)` at this point as its initialised. If it's anything other than 0 then we should revert.

0xmovses replied:
It is _technically_ `0x` at this point in the program, that is, when it is uninitialised, before it is set further down with the `recipient` arg, solidity does this by default.  (If it wasn't `0x` the tests would not pass). But we can remove this check as I guess its fairly redundant anyway and doesn't provide much value.

cde8 replied:
I'm sorry, Richard. It will always hold the address of the recipient because the bridge receives the recipient's address from the initiating event on the other side. It will never be 0x, see the abstract bridge code.

0xmovses replied:
The `recipient` will never be `0x` yes, but this check doesn't check `recipient` its checking against `transfer.recipient` which is the solidity struct field. Which is `0x` here.

0xmovses commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,81 @@
+
+        emit BridgeTransferAssetsLocked(bridgeTransferId, recipient, amount, hashLock, timeLock);
+
+        return true;
```
Do you mean the case where the same address might be swapping at the same time, as that recipient address is the key for the balances mapping?

What's the suggested change here? 

In the case of the same recipient address, we could use `bridgeTransferIds` to store the locked WETH instead of address, that's probably better in any case so I will add this.

0xmovses replied:
Makse sense! Ok

0xmovses replied:
The check now happens inside the `withdrawETH` function 
`if (poolBalance =< amount) revert ZeroAmount();`

0xmovses commented:
Ok using a simple pool now, and addressed your other points re. checks for WETH balance. Issue for a Vault contract has been created here and added to the feature branch checklist.  
https://github.com/movementlabsxyz/movement/issues/269

cde8 commented on protocol-units/bridge/contracts/src/AtomicBridgeCounterparty.sol:
```diff
@@ -0,0 +1,86 @@
+        address recipient,
+        uint256 amount
+    ) external onlyOwner returns (bool) {
+        if (amount == 0) revert ZeroAmount();
```
```suggestion
```

cde8 reviewed (approved):
Looking good Richard, lets merge.
